### PR TITLE
feat: add namespace to scheduler

### PR DESCRIPTION
**Pull request by aaronflorey**
Kept it simple by just adding a namespace to the scheduler interface and setting it in the engine. I added a parseCallback getter method so the user can get the callback and serialize it themselves, leaving the implementation up to them instead of you having to figure it out.

### CODE REVIEW

File: src/Core/Engine.php [hunk 1]
```diff
@@ -56,13 +56,17 @@ public function start(Run $run): void
     {
         $this->configure($run);
 
+        $this->scheduler->setNamespace($run->namespace);
```
Comment (ksassnowski) [inline src/Core/Engine.php:59]: This should go inside the `configure` method.

File: src/Core/Engine.php [hunk 2]
```diff
@@ -56,13 +56,17 @@ public function start(Run $run): void
...
+
         $this->eventDispatcher->dispatch(
             new RunStarting($run),
             RunStarting::NAME,
         );
 
-        foreach ($run->startRequests as $request) {
-            $this->scheduleRequest($request);
+        if ($this->scheduler->empty()) {
```
Comment (ksassnowski) [inline src/Core/Engine.php:66]: Can you explain why this change is necessary?

File: src/Core/RunFactory.php [hunk 3]
```diff
@@ -40,6 +40,7 @@ public function fromSpider(SpiderInterface $spider, ?Overrides $overrides = null
 
         return new Run(
             $spider->getInitialRequests(),
+            $spider::class,
```
Comment (ksassnowski) [inline src/Core/RunFactory.php:43]: We should also add a test to `RunFactoryTest.php` to check that the spider's class name gets passed as the default namespace.

**Review by ksassnowski — CHANGES_REQUESTED** [review]

File: src/Core/Engine.php (hunk 2 above)
Comment (aaronflorey) [reply src/Core/Engine.php:66]: I did this because with the Redis Scheduler, stopping and starting a run will continue on from where it left off.

So if the scheduler already has items in the queue, we should skip adding the initial requests onto the end of the scheduler.

Comment (ksassnowski) [reply src/Core/Engine.php:64]: I see. Then this should also be covered by tests, however.

File: tests/Core/EngineTest.php [hunk 4]
```diff
@@ -72,14 +72,41 @@ public function testCrawlsStartUrls(): void
...
+            $this->makeRequest('http://localhost:8000/test1'),
+            $this->makeRequest('http://localhost:8000/test2'),
+        ];
+
+        $run = new Run($startRequests, 'DefaultSpider');
+
+        $dispatcher = new EventDispatcher();
+        $scheduler = new ArrayRequestScheduler($this->clock);
+        $scheduler->schedule($this->makeRequest('http://localhost:8000/test3'));
+        $engine = new Engine(
```
Comment (ksassnowski) [inline tests/Core/EngineTest.php:95]: I think it would be easier if we made `ArrayRequestScheduler` a private field on the test as well. Then we won't have to duplicate the entire engine setup just for this test. We could instead just do this

```php
$this->scheduler->schedule($this->makeRequest('http://localhost:8000/test3'));
```

File: tests/Core/EngineTest.php [hunk 5]
```diff
@@ -72,14 +72,41 @@ public function testCrawlsStartUrls(): void
...
+            new Downloader(new Client(), $dispatcher),
+            new ItemPipeline($dispatcher),
+            new Processor($dispatcher),
+            $dispatcher,
+        );
+
+        $engine->start($run);
+
+        $this->assertRouteWasCrawledTimes('/test1', 0);
+        $this->assertRouteWasCrawledTimes('/test2', 0);
```
Comment (ksassnowski) [inline tests/Core/EngineTest.php:106]: These assertions are actually failing because of a bug in the `assertRouteWasCrawledTimes` method. The easiest fix would be to use `assertRouteWasNotCrawled` instead.

```suggestion
        $this->assertRouteWasNotCalled('/test1');
        $this->assertRouteWasNotCalled('/test2');
```

File: tests/Core/EngineTest.php [hunk 6]
```diff
@@ -72,14 +72,41 @@ public function testCrawlsStartUrls(): void
             $this->makeRequest('http://localhost:8000/test1'),
             $this->makeRequest('http://localhost:8000/test2'),
         ];
-        $run = new Run($startRequests);
+        $run = new Run($startRequests, 'DefaultSpider');
```
Comment (ksassnowski) [inline tests/Core/EngineTest.php:75]: If the actual value of a parameter in a test doesn't matter and we only care about what it represents, we use `::this-notation::`. This applies to all other places that pass the new namespace parameter, too.

```suggestion
        $run = new Run($startRequests, '::namespace::');
```

I've written a blog post on this topic in case you're interested in the rationale behind this: https://www.kai-sassnowski.com/post/reducing-unnecessary-details-in-tests/

**Review by ksassnowski — CHANGES_REQUESTED** [review]
There are few failing tests because they haven't been updated to provide the `$namespace` parameter to the `Run` class yet. Chances are you missed them because the integration tests don't get run locally by default.

To also run the integration tests:

1. Make sure `puppeteer` is installed: `npm install puppeteer` (just make sure to not commit any of the Javascript stuff. Could probably go inside `.gitignore`, too). 
2. Start the test server `composer test-server`
3. Run the test suite as normal

There shouldn't be any skipped tests when running the test suite in this way. 

There's a chance that `ExecuteJavascriptMiddlewareTest::testUpdateResponseBodyWithHtmlAfterExecutingJavascript` fails locally. You can ignore that as it runs correctly inside the GitHub action. I haven't been able to figure out what's going on with that test yet.

File: tests/Core/EngineTest.php (hunk 6 above)
Comment (aaronflorey) [reply tests/Core/EngineTest.php:75]: That actually makes a lot of sense, cheers!

File: tests/Core/EngineTest.php (hunk 5 above)
Comment (aaronflorey) [reply tests/Core/EngineTest.php:106]: sorry i wrote this while very tired, i'll look over things before committing again

**Review by ksassnowski — APPROVED** [review]

**Comment by ksassnowski** [comment]
Thanks for the contribution!

As a side note, I won't immediately release this as this is a breaking change and would require a major version bump. Before I do that I have a few other things I want to add first.
